microproxy/proxy.py

```diff
@@ -1,106 +1,19 @@
-                current_context = yield current_layer.process_and_return_context()
-                logger.debug("Leave {0} Layer".format(current_layer))
-                current_layer = self.next_layer(current_layer, current_context)
-            except gen.TimeoutError:
-                src_stream.close()
-                break
-            except DestNotConnectedError:
-                logger.debug("destination not conected")
-                break
-            except DestStreamClosedError:
-                logger.error("destination stream closed unexceptedly")
-                src_stream.close()
-                break
-            except SrcStreamClosedError:
-                logger.error("source stream closed unexceptedly")
-                break
-            except iostream.StreamClosedError:
-                logger.error("stream closed")
-                src_stream.close()
-                break
-            except Exception:
-                raise
-
-        raise gen.Return(None)
-
-    def get_first_layer(self, context):
-        mode = context.config["mode"]
-        if mode == "socks":
-            return SocksLayer(context)
-        elif mode == "transparent":
-            return TransparentLayer(context)
-        elif mode == "replay":
-            return ReplayLayer(context)
-        else:
-            raise ValueError("Unsupport proxy mode: {0}".format(mode))
-
-    def next_layer(self, current_layer, context):
-        http_ports = [80] + context.config["http_port"]
-        https_ports = [443] + context.config["https_port"]
-
-        if isinstance(current_layer, (SocksLayer, TransparentLayer)):
-            if context.port in http_ports:
-                context.scheme = "http"
-                return Http1Layer(context)
-            elif context.port in https_ports:
-                return TlsLayer(context, self.cert_store)
-            else:
-                return ForwardLayer(context)
-
-        if isinstance(current_layer, TlsLayer):
-            if context.scheme == "https":
-                return Http1Layer(context)
-            elif context.scheme == "h2":
-                return Http2Layer(context)
-            else:
-                return ForwardLayer(context)
-
-        if isinstance(current_layer, ReplayLayer):
-            if context.scheme in ("http", "https"):
-                return Http1Layer(context)
-            elif context.scheme == "h2":
-                return Http2Layer(context)
-            else:
-                return ForwardLayer(context)
-
-
 class ProxyServer(tcpserver.TCPServer):
     def __init__(self, config, **kwargs):
         super(ProxyServer, self).__init__(**kwargs)
         self.config = config
-        self.interceptor = Interceptor(config=config)
-        self.layer_manager = LayerManager(config)
-        self.event_manager = EventManager(config, self)
 
     def _handle_connection(self, connection, address):
         try:
```

Comment:
Maybe we should add a note to describe that this method override TCPServer's method.
Because I remembered that I had removed address argument last time that I thought it didn't be used but it was used by tornado TCPServer.